Feature/tdr 8/release to npm

oatymart opened the pull request:
https://oat-sa.atlassian.net/browse/TDR-8

- created new command `taoRelease npmRelease` & adapted existing code
- github release is working, both for traditional extension repos and npm package repos
- `npm publish` step is left as a manual step, due to the complexity of validating the user's authentication to publish to npmjs.com
- unit tests updated (not full coverage)
<img width="604" alt="Screen Shot 2020-01-06 at 11 13 49" src="https://user-images.githubusercontent.com/43652944/71811692-aa054380-3075-11ea-9ee4-64d4ef438e0b.png">

To test the `npmRelease`, I have been using a dummy package (https://github.com/oatymart/tao-foobaz-fe) and local Verdaccio server to avoid creating side effects on public registries.

krampstudio commented on src/release.js:
```diff
@@ -58,9 +60,32 @@ module.exports = function taoExtensionReleaseFactory(params = {}) {
+        /**
+         * Run the build command of the package so we release latest build
+         */
+        async buildPackage() {
```
each package is supposed to build itself when running publish, this is the role of the pacakge's `prepare` action, like  https://github.com/oat-sa/tao-item-runner-fe/blob/master/package.json#L20
the `prepare` script is called after install and before publish (see https://docs.npmjs.com/misc/scripts#description) it's an official step of the npm lifecycle, when `build` is an internal convention

oatymart replied:
Ok, understood, but this function is run as part of the github release phase. It's the equivalent of the `grunt bundle` command which is run on traditional extensions. It will guarantee that the package-lock.json is updated to match the package.json, catch unbuildable code, and generate any artifacts which need to be committed to the repo at release time.

krampstudio replied:
For an npm package this is not necessary otherwise it means the package cannot be published in its current state. This is due to the difference between the github repo (the source) and the npm package (built runtime/artifact) and the transition between the source state to the artifact state is `npm install` and `npm publish`

krampstudio commented on src/release.js:
```diff
@@ -22,6 +22,7 @@
  * @author Anton Tsymuk <anton@taotesting.com>
  */
 
+const fs = require('fs');
```
I have the feeling the code isn't well organized. We are mixing the code to release a tao extention and an npm package. When we will add other types (composer libs for example) the code will become complicated and more difficult to maintain.
What about splitting in 3 files : release.js (general / adaptor), release/npm and release/tao
And when you call a method on `release` it can delegate the call to the correct implementation (adapter pattern).

krampstudio replied:
```js
module.exports = function releaseFactory(params = {}) {
  
  //determine the subject upfront
  const releaseAdaptee = require(`./release/${subject}.js`)(params);

  return {
    someGenericMethod() {
        //code is here
    },
    someSubjectDependentMethod(){
       return releaseAdaptee.someSubjectDependentMethod();
    }
  }
};
```

oatymart replied:
The idea seems good.
I don't know if I see the need to make the Adaptee a factory taking the exact same params as its parent. I would propose to do it as a plain object containing 4 or 5 methods which are called by the releaseFactory. A few injected dependencies and return values should be all I need.

However, after sleeping on the idea I may change my mind!

krampstudio commented on tests/unit/npmPackage/npmCommand/test.js:
```diff
@@ -0,0 +1,76 @@
+
+const proxyquire = require('proxyquire');
+const sinon = require('sinon');
+// const test = require('tape');
```
forgotten comment

krampstudio commented on src/npmPackage.js:
```diff
@@ -0,0 +1,150 @@
+         * Run `npm publish` command
+         * @returns {Promise}
+         */
+        publish(dryRun = true, myRegistry = true) {
```
dry run should be false by default

krampstudio commented on src/npmPackage.js:
```diff
@@ -0,0 +1,150 @@
+        publish(dryRun = true, myRegistry = true) {
+            // Flags for testing purposes:
+            const dryRunFlag = '--dry-run';
+            const registryFlag = '--registry http://localhost:4873';
```
maybe change the parameter `myRegistry` to `(registry = 'https://registry.npmjs.org')` ?

oatymart replied:
Done.

oatymart commented:
@krampstudio I've completed the restructuring of extension- and package-specific code from `release.js` into 2 files, `release/extensionApi.js` and `release/packageApi.js`.

The integration tests are currently passing and I've tested all the commands on real releases as well. The work that remains is to adapt a good number of unit tests.

oatymart commented on src/npmPackage.js:
```diff
@@ -0,0 +1,150 @@
+
+                const spawned = crossSpawn('npm', command.split(' '), opts);
+                spawned.on('close', code => {
+                    code === 0 ? resolve() : reject();
```
Ok, I added errors.

oatymart commented:
@krampstudio @jsconan The latest round of refactoring to Adapter/Adaptee pattern is ready for review. The unit tests are fixed and I've done a couple of test releases.

oatymart commented on src/release.js:
```diff
@@ -50,17 +51,105 @@ const taoInstanceFactory = require('./taoInstance.js');
+            else if (subjectType === 'package') {
+                adaptee = packageApi(params, data);
+            }
+        },
```
Ok that works for me.

oatymart commented:
Fixed the above comments and implemented a second PR checklist, which you can see in action here: https://github.com/oatymart/tao-foobaz-fe/pull/25

If you prefer no checklist, that can also be easily done.
